criu/sk-inet.c

```diff
@@ -804,6 +812,44 @@ static int open_inet_sk(struct file_desc *d, int *new_fd)
 	if (set_netns(ie->ns_id))
 		return -1;
 
+#ifdef CONFIG_HAS_SELINUX
+	/*
+	 * When running in a SELinux environment processes and threads will
+	 * be labeled with the same process labels they had during checkpoint.
+	 * The process labels are set as late as possible during restore.
+	 * Sockets created during restore, however, will be labeled with
+	 * CRIU's context during restore. These sockets might not work
+	 * correctly once the process is restored and switched to another
+	 * context.
+	 *
+	 * This tells SELinux to label sockets with the appropriate label.
+	 *
+	 * As with all SELinux related labeling, all processes and sockets
+	 * are labeled with the same SELinux label.
+	 */
+	if (kdat.lsm == LSMTYPE__SELINUX) {
+		/* Use the SELinux label from the first process/thread */
+		if (current->core[0]->thread_core &&
+		    current->core[0]->thread_core->creds &&
+		    current->core[0]->thread_core->creds->lsm_profile) {
+			/*
+			 * thread_core, thread_core->creds and
+			 * thread_core->creds->lsm_profile are all optional
+			 * protobuf fields. This makes sure these exist
+			 * before accessing them.
+			 */
+			int ret = setsockcreatecon_raw(current->core[0]->
```

Comment:
A socket label can be different than a process label, can't it? Should we use fgetxattr(..., "security.selinux") to dump a socket label?

Reply:
Yes, a socket can have a different label. This is the minimal fix for the container (Podman) use case where everything in the container has the same label. It is similar to the current handling of threads. CRIU currently restores all threads with the same SELinux label. In theory threads could have different process labels (if those labels are bounded).

This fix is still needed as we need to make sure that sockets created in the future (after restore) have at least the label of the process.

In a follow up patch I can read and store all socket labels and re-create the correctly on restore.

The correct fix would also require checking if the user has used setsockcreatecon*() and restore that value as well.

If possible I would like to get this into CRIU and then the follow-up would read out /proc/PID/attr/sockcreate to see if something is set and store that for newly created sockets and read out the socket label to restore it (fgetxattr).

Reply:
> Yes, a socket can have a different label. This is the minimal fix for the container (Podman) use case where everything in the container has the same label. It is similar to the current handling of threads. CRIU currently restores all threads with the same SELinux label. In theory threads could have different process labels (if those labels are bounded).

I'm ok with this, but we have to be sure that we will not restore a wrong label for a socket. For this we need to dump a socket label and then on restore we need to check that a socket label and a process label are the same.